cura/Settings/MachineManager.py

```diff
@@ -1496,3 +783,268 @@ def setSettingForAllExtruders(self, setting_name: str, property_name: str, prope
+            for position in self._global_container_stack.extruders:
+                if position not in self._current_root_material_name:
+                    material = self._global_container_stack.extruders[position].material
+                    self._current_root_material_name[position] = material.getName()
+        return self._current_root_material_name
+
+    ##  Return the variant names in the extruder stack(s).
+    ##  For the variant in the global stack, use activeVariantBuildplateName
+    @pyqtProperty("QVariant", notify = activeVariantChanged)
+    def activeVariantNames(self):
+        result = {}
+
+        active_stacks = ExtruderManager.getInstance().getActiveExtruderStacks()
+        if active_stacks is not None:
+            for stack in active_stacks:
+                variant_container = stack.variant
+                position = stack.getMetaDataEntry("position")
+                if variant_container and variant_container != self._empty_variant_container:
+                    result[position] = variant_container.getName()
+
+        return result
+
+    def _setEmptyQuality(self):
+        self._current_quality_group = None
+        self._current_quality_changes_group = None
+        self._global_container_stack.quality = self._empty_quality_container
+        self._global_container_stack.qualityChanges = self._empty_quality_changes_container
+        for extruder in self._global_container_stack.extruders.values():
+            extruder.quality = self._empty_quality_container
+            extruder.qualityChanges = self._empty_quality_changes_container
+
+        self.activeQualityGroupChanged.emit()
+        self.activeQualityChangesGroupChanged.emit()
+
+    def _setQualityGroup(self, quality_group, empty_quality_changes = True):
+        self._current_quality_group = quality_group
+        if empty_quality_changes:
+            self._current_quality_changes_group = None
+
+        # Set quality and quality_changes for the GlobalStack
+        self._global_container_stack.quality = quality_group.node_for_global.getContainer()
+        if empty_quality_changes:
+            self._global_container_stack.qualityChanges = self._empty_quality_changes_container
+
+        # Set quality and quality_changes for each ExtruderStack
+        for position, node in quality_group.nodes_for_extruders.items():
+            self._global_container_stack.extruders[position].quality = node.getContainer()
+            if empty_quality_changes:
+                self._global_container_stack.extruders[position].qualityChanges = self._empty_quality_changes_container
+
+        self.activeQualityGroupChanged.emit()
+        self.activeQualityChangesGroupChanged.emit()
+
+    def _setQualityChangesGroup(self, quality_changes_group):
+        # TODO: quality_changes groups depend on a quality_type. Here it's fetching the quality_types every time.
+        #       Can we do this better, like caching the quality group a quality_changes group depends on?
```

Comment:
You missed a TODO here.

Reply:
Done in a220b5585.

Reply:
So it looks like the caching was impossible?